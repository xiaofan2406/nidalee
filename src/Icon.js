/* @flow */
import * as React from 'react';
import styled, { cx } from 'react-emotion';

export const I = styled.i`
  font-size: ${({ size }) => size}px;
  color: ${({ color }) => color};
  margin: 0px 2px;
`;

const sizeMap = size =>
  ({ small: 14, regular: 16, large: 20 }: {
    [key: Size]: number,
  })[size];

const Icon = ({ color, size, type, name, className, ...rest }: IconProps) => (
  <I
    color={color}
    size={sizeMap(size)}
    className={cx([name && type && `fa${type[0]} fa-${name}`, className])}
    {...rest}
  />
);

Icon.defaultProps = {
  type: 'regular',
  name: '',
  size: 'regular',
};

export default Icon;
