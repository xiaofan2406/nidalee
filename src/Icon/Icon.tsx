import React, { FC } from 'react';
import { css } from '@emotion/core';
import { theme } from '../styles';

export interface IconProps extends React.HTMLAttributes<HTMLElement> {
  type?: 'solid' | 'regular' | 'light' | 'brands';
  name?: string;
  className?: string;
  size?: 'small' | 'regular' | 'large';
  color?: string;
}

const sizeMap = (size: IconProps['size']) =>
  ({ small: 12, regular: 16, large: 24 }[size!]);

const Icon: FC<IconProps> = ({
  color,
  size,
  type,
  name,
  className,
  ...rest
}) => {
  // TODO assert on 1 type of usage: className or type&name
  return (
    <i
      css={css`
        margin: 0 2px;
        color: ${color};
        font-size: ${sizeMap(size!)}px;
      `}
      className={
        className ||
        [name && type && `fa${type[0]}`, name && type && `fa-${name}`]
          .filter(Boolean)
          .join(' ')
          .trim()
      }
      {...rest}
    />
  );
};

Icon.defaultProps = {
  type: 'regular',
  size: 'regular',
  color: theme.primaryColor,
};

export default Icon;
