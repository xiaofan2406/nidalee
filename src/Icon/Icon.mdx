---
name: Icon
---

import { Playground, PropsTable } from 'docz';
import Icon from './Icon';

# Icon

> In order to use Icon component, you will need to manually add font awesome assets.

## Props

<PropsTable of={Icon} />

## Basic usage

<Playground>
  <Icon type="solid" name="address-book" />
  <Icon type="regular" name="address-book" />
  <Icon type="brands" name="accessible-icon" />
  <Icon className="fas fa-adjust" />
</Playground>
