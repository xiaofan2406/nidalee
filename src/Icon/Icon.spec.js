import React from 'react';
import { render } from 'react-testing-library';
import Icon from './Icon';

test('snapshot matches with type and name', () => {
  const { container } = render(<Icon type="solid" name="arrow-up" />);
  expect(container.firstChild).toMatchSnapshot();
});

test('snapshot matches with className', () => {
  const { container } = render(<Icon className="fas fa-arrow-up" />);
  expect(container.firstChild).toMatchSnapshot();
});
