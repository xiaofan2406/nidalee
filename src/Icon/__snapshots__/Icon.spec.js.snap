// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches with className 1`] = `
<i
  class="fas fa-arrow-up css-13h5fv5-Icon"
/>
`;

exports[`snapshot matches with type and name 1`] = `
<i
  class="fas fa-arrow-up css-13h5fv5-Icon"
/>
`;
